### DISCUSSION TITLE
class methods not returning data

### DISCUSSION THREAD

**Discussion by mykolachub** [HowProgrammingWorks/LiveQA#167] [Q&A]
### Идея:

При создании инстанса класса `Foo` метод `init` в конструкторе должен проходиться по коллекции `types` и запускать соответствующий хендлер и возвращать данные. 

### Проблема:

Создав инстанс класса, метод `init` запускает необходимые хендлер, однако они не возвращают данные. В чем моя ошибка и как можно иначе решить подобную проблему?

```js
'use strict';

class Foo {
  constructor(type) {
    this.type = type;
    this.init();
  }

  init() {
    const types = {
      bar: this.handleBar,
      baz: this.handleBaz,
    };
    const handler = types[this.type];
    return handler();
  }

  handleBar() {
    console.log('calling this.handleBar method');
    return 'Bar was handled';
  }

  handleBaz() {
    console.log('calling this.handleBaz method');
    return 'Baz was handled';
  }
}

const bar = new Foo('bar');
console.dir({ bar }); // expected 'bar' to be 'Bar was handled'

/*

output

> calling this.handleBar method
> { bar: Foo { type: 'bar' } }

*/

```

**Comment by danzelito** [comment]
const types = {
      bar: this.handleBar.bind(this),
      baz: this.handleBaz.bind(this),
    };
Так напиши и будет тебе счастье. Это если я правильно в твоей цели разобрался. Если не понял почему -  напиши, объясню.

**Reply by mykolachub** [reply]
Моя задача была добиться того, чтобы при создании инстанса класса, он сразу же возвращался результат автоматически обработанный `init`. 
Но как я уже понял, в чистом виде этого добиться нельзя, так как `constructor` возвращает всегда объект. Но можно переопределить поведение:
```js
// ...
constructor(type) {
    this.type = type;
    return () => this.init();
  }
// ...

const bar = new Foo('bar')();
```
 `constructor` возвращает ссылку на метод, который при инстансе нужно будет вручную вызвать, что правда, тоже не решает изначальную проблему.
